### PR TITLE
[Rust] Lance File Writer

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/datatypes.rs [hunk 1]
```diff
@@ -313,9 +301,13 @@ impl Field {
...
     fn max_id(&self) -> i32 {
-        min(
+        max(
             self.id,
-            self.children.iter().map(|c| c.id).min().unwrap_or(i32::MAX),
+            self.children
+                .iter()
+                .map(|c| c.max_id())
+                .max()
+                .unwrap_or(i32::MIN),
```
Comment (changhiskhan) [inline rust/src/datatypes.rs:310]: nbd but i thought I saw elsewhere `-1` was the place holder? Is it too cumbersome to use `Option<u32>` as field/parent id?

File: rust/src/datatypes.rs [hunk 2]
```diff
@@ -420,8 +412,8 @@ impl TryFrom<&ArrowField> for Field {
             name: field.name().clone(),
             logical_type: LogicalType::try_from(field.data_type())?,
             encoding: match field.data_type() {
-                dt if is_numeric(dt) => Some(Encoding::Plain),
-                dt if is_binary(dt) => Some(Encoding::VarBinary),
+                dt if dt.is_numeric() || matches!(dt, DataType::Boolean) => Some(Encoding::Plain),
```
Comment (changhiskhan) [inline rust/src/datatypes.rs:415]: maybe move `is_fixed_stride` from datatypes to arrow.rs and use that here?

File: rust/src/io.rs [hunk 3]
```diff
@@ -29,13 +29,13 @@ pub mod object_reader;
 pub mod object_store;
 pub mod object_writer;
 pub mod reader;
+mod writer;
```
Comment (changhiskhan) [inline rust/src/io.rs:32]: should this be pub mod?

File: rust/src/io/object_writer.rs [hunk 4]
```diff
@@ -15,29 +15,41 @@
...
+
     // TODO: wrap writer with a BufWriter.
     #[pin]
-    writer: &'a mut (dyn AsyncWrite + Unpin + Send),
+    writer: Box<dyn AsyncWrite + Unpin + Send>,
+    multipart_id: MultipartId,
     cursor: usize,
 }
 
-impl<'a> ObjectWriter<'a> {
```
Comment (changhiskhan) [inline rust/src/io/object_writer.rs:38]: curious why you got rid of all of the lifetimes

File: rust/src/format/page_table.rs [hunk 5]
```diff
@@ -68,7 +82,47 @@ impl PageTable {
         Ok(Self { pages })
     }
 
-    pub fn set_page_info(&mut self) {}
+    pub async fn write(&self, writer: &mut ObjectWriter) -> Result<usize> {
+        let pos = writer.tell();
+        assert!(!self.pages.is_empty());
+        let num_columns = self.pages.keys().max().unwrap() + 1;
+        let num_batches = self
```
Comment (changhiskhan) [inline rust/src/format/page_table.rs:89]: page_table.num_columns and page_table.num_batches?

File: rust/src/format/page_table.rs [hunk 6]
```diff
@@ -15,42 +15,56 @@
...
+    pub fn new(position: usize, length: usize) -> Self {
+        Self { position, length }
+    }
+}
+
+/// Page lookup table.
 #[derive(Debug, Default)]
 pub struct PageTable {
-    pages: HashMap<i32, HashMap<i32, PageInfo>>,
+    pages: BTreeMap<i32, BTreeMap<i32, PageInfo>>,
```
Comment (changhiskhan) [inline rust/src/format/page_table.rs:45]: add a comment here to clarify this is  { field_id -> { batch_id -> page_info}} ?

File: rust/src/io/writer.rs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        }
+    }
+
+    /// Write a [RecordBatch] to the open file.
+    ///
+    /// Returns [Err] if the schema does not match with the batch.
+    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        for field in &self.schema.fields {
+            let column_id = batch.schema().index_of(&field.name)?;
+            let array = batch.column(column_id);
```
Comment (changhiskhan) [inline rust/src/io/writer.rs:59]: should we just implement `RecordBatch.column(&str)` ?

**Review by changhiskhan — COMMENTED** [review]
A few questions

File: rust/src/io/writer.rs (hunk 7 above)
Comment (eddyxu) [reply rust/src/io/writer.rs:59]: You mean extend it? sure, why not :) 

maybe we should have a `lance::preclude::*`

File: rust/src/datatypes.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/datatypes.rs:310]: Oh, right, good catch. fixed.

File: rust/src/datatypes.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/datatypes.rs:415]: done.

File: rust/src/io/object_writer.rs (hunk 4 above)
Comment (eddyxu) [reply rust/src/io/object_writer.rs:38]: Seems easier to "create an object" with the internal reference as copy  and pass around . 

could revisit it if the overhead is too big.

File: rust/src/format/page_table.rs (hunk 5 above)
Comment (eddyxu) [reply rust/src/format/page_table.rs:89]: not sure wdym?

PageTable does not have these two fields.

File: rust/src/format/page_table.rs (hunk 6 above)
Comment (eddyxu) [reply rust/src/format/page_table.rs:45]: done

File: rust/src/io.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/io.rs:32]: was thining just re-export writer/reader via io, and make writer/reader to be private.

it takes quite some typing to import `lance::io::reader::FileReader` ? wdyt

File: rust/src/io/writer.rs (hunk 7 above)
Comment (eddyxu) [reply rust/src/io/writer.rs:59]: lemme do it in the follow up PRs?

File: rust/src/io.rs (hunk 3 above)
Comment (changhiskhan) [reply rust/src/io.rs:32]: fine by me. was just curious why reader and writer were different

**Review by changhiskhan — APPROVED** [review]
some CI error due to cargo fmt

File: rust/src/io.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/io.rs:32]: was thinking to make `reader` the same as `writer` later.

**Pull request merged**